### ISSUE TITLE
LARGE uploads to S3 "works" but server says:  NoSuchUpload...

### ISSUE THREAD

**Issue by kjetilge** [CollectionFS/Meteor-CollectionFS#254]
Wen the server is finished with READ CHUNK it ends with:

W20140403-23:24:34.736(2)? (STDERR) NoSuchUpload: The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
W20140403-23:24:34.737(2)? (STDERR)     at Request.extractError (/Users/kjetil/.meteorite/packages/cfs-s3/CollectionFS/Meteor-cfs-s3/8db294ab28729e9dd8d49daaa7d8fdf144abda1a/.build/npm/node_modules/aws-sdk/lib/services/s3.js:235:35)

When I check the bucket, the 1,7 GB file is there! (using chunkSize 320).
This only happens when uploading large video files.

**Comment by 3thanZ** [comment]
Hi,

Has anyone had any luck at all uploading files greater than 5MB onto S3 using cfs:s3? As of 10 Jan 2015, I'm still encountering this issue.

Any help will be very much appreciated.

Server side:
```javascript
FS.debug = true;

var fileStore = new FS.Store.S3("files", {
	// region: 'ap-southeast-1',
	accessKeyId: 'xxx',
	secretAccessKey: 'xxx',
	bucket: 'files'
});

Uploads = new FS.Collection("files", {
	stores: [fileStore],
	chunkSize: 2048000
});

Meteor.startup(function() {
	Uploads.allow({
		insert: function(userId, doc){
			return true;
		},
		update: function(userId, doc, fields, modifier) {
			return true;
		},
		remove: function (userId, doc){
			return true;
		},
		download: function () {
			return true;
		}
	});
});
```

Client side:
```javascript
var fileStore = new FS.Store.S3("files");

Uploads = new FS.Collection("files", {
	stores: [fileStore],
	chunkSize: 2048000
});
```

With FS.debug = true, the server side console looks like this:
```
I20150110-04:55:32.064(8)? FileWorker ADDED - calling deleteChunks for JzQGCRxeyaD7gP6ha
I20150110-04:55:32.067(8)? _tempstore FileSystem mounted on: /Users/ethans/Desktop/dk/dk-atlas-web/trunk/src/.meteor/local/cfs/files/_tempstore
I20150110-04:55:32.067(8)? TempStore is mounted on storage.filesystem
I20150110-04:55:32.079(8)? FS.HTTP.unmount:
I20150110-04:55:32.080(8)? {}
I20150110-04:55:32.081(8)? Registered HTTP method URLs:
I20150110-04:55:32.081(8)? /cfs/files/:collectionName/:id/:filename
I20150110-04:55:32.081(8)? /cfs/files/:collectionName/:id
I20150110-04:55:32.081(8)? /cfs/files/:collectionName
=> Meteor server restarted
I20150110-04:56:13.074(8)? token: eyJhdXRoVG9rZW4iOiJ1aV9Fd3FTVEpjeFNCbC1IY25BSlBBYndPV0JCTTRyZE41N3lsZ3plSWZ0In0=
I20150110-04:56:13.076(8)? HTTP PUT (update) handler received chunk:  0
I20150110-04:56:13.077(8)? GET FILERECORD: XF5riet3EdZ3Zwg8u
I20150110-04:56:13.081(8)? createWriteStream _tempstore, internal: true
I20150110-04:56:13.082(8)? createWriteStreamForFileKey _tempstore
I20150110-04:56:13.094(8)? -----------FINISH STREAM _tempstore
I20150110-04:56:13.095(8)? SA FileSystem - DONE!! fileKey: "files-XF5riet3EdZ3Zwg8u-0.chunk"
I20150110-04:56:13.096(8)? -----------STORED STREAM _tempstore
I20150110-04:56:13.096(8)? -----------CLOSE STREAM _tempstore
I20150110-04:56:13.098(8)? TempStore progress: Received 1 of 5 chunks for Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:13.098(8)? UPDATE: {"$set":{"chunkCount":1}}
I20150110-04:56:13.567(8)? token: eyJhdXRoVG9rZW4iOiJ1aV9Fd3FTVEpjeFNCbC1IY25BSlBBYndPV0JCTTRyZE41N3lsZ3plSWZ0In0=
I20150110-04:56:13.574(8)? HTTP PUT (update) handler received chunk:  1
I20150110-04:56:13.575(8)? GET FILERECORD: XF5riet3EdZ3Zwg8u
I20150110-04:56:13.584(8)? createWriteStream _tempstore, internal: true
I20150110-04:56:13.585(8)? createWriteStreamForFileKey _tempstore
I20150110-04:56:13.597(8)? -----------FINISH STREAM _tempstore
I20150110-04:56:13.598(8)? SA FileSystem - DONE!! fileKey: "files-XF5riet3EdZ3Zwg8u-1.chunk"
I20150110-04:56:13.598(8)? -----------STORED STREAM _tempstore
I20150110-04:56:13.599(8)? -----------CLOSE STREAM _tempstore
I20150110-04:56:13.602(8)? TempStore progress: Received 2 of 5 chunks for Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:13.602(8)? UPDATE: {"$set":{"chunkCount":2}}
I20150110-04:56:14.075(8)? token: eyJhdXRoVG9rZW4iOiJ1aV9Fd3FTVEpjeFNCbC1IY25BSlBBYndPV0JCTTRyZE41N3lsZ3plSWZ0In0=
I20150110-04:56:14.077(8)? HTTP PUT (update) handler received chunk:  2
I20150110-04:56:14.078(8)? GET FILERECORD: XF5riet3EdZ3Zwg8u
I20150110-04:56:14.082(8)? createWriteStream _tempstore, internal: true
I20150110-04:56:14.083(8)? createWriteStreamForFileKey _tempstore
I20150110-04:56:14.090(8)? -----------FINISH STREAM _tempstore
I20150110-04:56:14.092(8)? SA FileSystem - DONE!! fileKey: "files-XF5riet3EdZ3Zwg8u-2.chunk"
I20150110-04:56:14.092(8)? -----------STORED STREAM _tempstore
I20150110-04:56:14.092(8)? -----------CLOSE STREAM _tempstore
I20150110-04:56:14.097(8)? TempStore progress: Received 3 of 5 chunks for Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:14.098(8)? UPDATE: {"$set":{"chunkCount":3}}
I20150110-04:56:14.612(8)? token: eyJhdXRoVG9rZW4iOiJ1aV9Fd3FTVEpjeFNCbC1IY25BSlBBYndPV0JCTTRyZE41N3lsZ3plSWZ0In0=
I20150110-04:56:14.614(8)? HTTP PUT (update) handler received chunk:  3
I20150110-04:56:14.614(8)? GET FILERECORD: XF5riet3EdZ3Zwg8u
I20150110-04:56:14.616(8)? createWriteStream _tempstore, internal: true
I20150110-04:56:14.617(8)? createWriteStreamForFileKey _tempstore
I20150110-04:56:14.624(8)? -----------FINISH STREAM _tempstore
I20150110-04:56:14.630(8)? SA FileSystem - DONE!! fileKey: "files-XF5riet3EdZ3Zwg8u-3.chunk"
I20150110-04:56:14.632(8)? -----------STORED STREAM _tempstore
I20150110-04:56:14.632(8)? -----------CLOSE STREAM _tempstore
I20150110-04:56:14.633(8)? TempStore progress: Received 4 of 5 chunks for Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:14.633(8)? UPDATE: {"$set":{"chunkCount":4}}
I20150110-04:56:15.075(8)? token: eyJhdXRoVG9rZW4iOiJ1aV9Fd3FTVEpjeFNCbC1IY25BSlBBYndPV0JCTTRyZE41N3lsZ3plSWZ0In0=
I20150110-04:56:15.078(8)? HTTP PUT (update) handler received chunk:  4
I20150110-04:56:15.078(8)? GET FILERECORD: XF5riet3EdZ3Zwg8u
I20150110-04:56:15.080(8)? createWriteStream _tempstore, internal: true
I20150110-04:56:15.081(8)? createWriteStreamForFileKey _tempstore
I20150110-04:56:15.086(8)? -----------FINISH STREAM _tempstore
I20150110-04:56:15.087(8)? SA FileSystem - DONE!! fileKey: "files-XF5riet3EdZ3Zwg8u-4.chunk"
I20150110-04:56:15.087(8)? -----------STORED STREAM _tempstore
I20150110-04:56:15.088(8)? -----------CLOSE STREAM _tempstore
I20150110-04:56:15.092(8)? TempStore progress: Received 5 of 5 chunks for Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:15.092(8)? UPDATE: {"$set":{"uploadedAt":"2015-01-09T20:56:15.092Z"},"$unset":{"chunkCount":1,"chunkSum":1,"chunkSize":1}}
I20150110-04:56:15.093(8)? Kevin Pool - Tibco - Big Fast Operational Data.pdf was successfully uploaded. You are seeing this informational message because you enabled debugging and you have not defined any listeners for the "uploaded" event on the files collection.
I20150110-04:56:15.099(8)? FileWorker ADDED - calling saveCopy files for XF5riet3EdZ3Zwg8u
I20150110-04:56:15.100(8)? saving to store files
I20150110-04:56:15.100(8)? createWriteStream files, internal: false
I20150110-04:56:15.100(8)? createWriteStreamForFileKey files
I20150110-04:56:15.135(8)? FS.TempStore creating read stream for XF5riet3EdZ3Zwg8u
I20150110-04:56:15.139(8)? createReadStream _tempstore
I20150110-04:56:15.140(8)? createReadStreamForFileKey _tempstore
I20150110-04:56:15.173(8)? createReadStream _tempstore
I20150110-04:56:15.173(8)? createReadStreamForFileKey _tempstore
I20150110-04:56:15.177(8)? createReadStream _tempstore
I20150110-04:56:15.178(8)? createReadStreamForFileKey _tempstore
I20150110-04:56:15.182(8)? createReadStream _tempstore
I20150110-04:56:15.182(8)? createReadStreamForFileKey _tempstore
I20150110-04:56:15.196(8)? createReadStream _tempstore
I20150110-04:56:15.197(8)? createReadStreamForFileKey _tempstore
I20150110-04:56:19.491(8)? -----------FINISH STREAM files
I20150110-04:56:19.492(8)? -----------FINISH STREAM files
I20150110-04:56:19.646(8)? SA S3 - DONE!!
I20150110-04:56:19.646(8)? -----------STORED STREAM files
I20150110-04:56:19.646(8)? -----------STORED STREAM files
I20150110-04:56:19.646(8)? SA files stored files/XF5riet3EdZ3Zwg8u-Kevin Pool - Tibco - Big Fast Operational Data.pdf
I20150110-04:56:19.647(8)? FS.File._saveChanges: files
I20150110-04:56:19.647(8)? UPDATE: {"$set":{"copies.files":{"name":"Kevin Pool - Tibco - Big Fast Operational Data.pdf","type":"application/pdf","size":5275648,"key":"files/XF5riet3EdZ3Zwg8u-Kevin Pool - Tibco - Big Fast Operational Data.pdf","updatedAt":"2015-01-09T20:56:19.645Z","createdAt":"2015-01-09T20:56:19.645Z"}}}
I20150110-04:56:19.648(8)? Kevin Pool - Tibco - Big Fast Operational Data.pdf was successfully saved to the files store. You are seeing this informational message because you enabled debugging and you have not defined any listeners for the "stored" event on the files collection.
I20150110-04:56:19.652(8)? FileWorker ADDED - calling deleteChunks for XF5riet3EdZ3Zwg8u
I20150110-04:56:19.653(8)? ---SA REMOVE
I20150110-04:56:19.653(8)? ---SA REMOVE
I20150110-04:56:19.654(8)? ---SA REMOVE
I20150110-04:56:19.654(8)? ---SA REMOVE
I20150110-04:56:19.654(8)? ---SA REMOVE
I20150110-04:56:22.659(8)? SA S3 - ERROR!!
I20150110-04:56:22.661(8)? -----------ERROR STREAM files The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
I20150110-04:56:22.661(8)? -----------ERROR STREAM files The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
W20150110-04:56:22.661(8)? (STDERR) 
W20150110-04:56:22.662(8)? (STDERR) /Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/request.js:32
W20150110-04:56:22.662(8)? (STDERR)           throw err;
W20150110-04:56:22.662(8)? (STDERR)                 ^
W20150110-04:56:22.676(8)? (STDERR) Error: Error storing file to the files store: The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
W20150110-04:56:22.677(8)? (STDERR)     at [object Object].<anonymous> (packages/cfs:collection/common.js:88:1)
W20150110-04:56:22.677(8)? (STDERR)     at [object Object].emit (events.js:106:17)
W20150110-04:56:22.677(8)? (STDERR)     at Writable.<anonymous> (packages/cfs:storage-adapter/storageAdapter.server.js:203:1)
W20150110-04:56:22.678(8)? (STDERR)     at Writable.emit (events.js:117:20)
W20150110-04:56:22.678(8)? (STDERR)     at Response.<anonymous> (packages/cfs:s3/s3.upload.stream2.js:152:1)
W20150110-04:56:22.678(8)? (STDERR)     at Request.<anonymous> (/Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/request.js:350:18)
W20150110-04:56:22.678(8)? (STDERR)     at Request.callListeners (/Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/sequential_executor.js:100:18)
W20150110-04:56:22.679(8)? (STDERR)     at Request.emit (/Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
W20150110-04:56:22.679(8)? (STDERR)     at Request.emit (/Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/request.js:604:14)
W20150110-04:56:22.679(8)? (STDERR)     at Request.transition (/Users/ethans/.meteor/packages/cfs_s3/.0.1.1.1lql3wn++os+web.browser+web.cordova/npm/node_modules/aws-sdk/lib/request.js:21:12)
```

**Comment by zda** [comment]
I am also seeing the same issue (100+ MB file and cfs:s3). Investigating.

**Comment by zda** [comment]
@aldeed, could you point me in the direction of where this error message is generated? I couldn't find it searching the various repos.

`Error storing file to the <storename> store: The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.`

**Comment by zda** [comment]
Update: I couldn't fix it by checking out previous versions of my app (even though I have reason to believe large uploads worked last month), by downgrading various packages and dependencies, or by forking `cfs:s3` to depend on the latest aws-sdk.

**Comment by Nlammertyn** [comment]
I'm having the exact same issue. I was thinking there might be a mistake with the incrementation of the partNumber or similar and that it tries to upload an extra chunk after the last one. Tried to fiddle with it, but alas no solution. I might dive into the code a bit more, but in the meantime any help/insights would be awesome.

**Comment by raix** [comment]
@Nlammertyn The s3 storage adapter got 3 different implementations, its currently using the "correct" streaming method - but theres a chunk limit at 5mb on s3 uploads so I'm thinking theres an issue with the partial upload.

We could switch to one of the other solutions (comments in the package.js) but I havent got time atm to look in to it.

Any attempts on debugging this would be awesome - if anybody feels up to it it would be nice to have some unit test/qa test for storage adapters in general.

**Comment by Nlammertyn** [comment]
@raix Thanks for the reply, I'll test out the other methods tomorrow. If one of them works I'll use that for now and I'll take a look at the stream2 implementation as I do agree that method is ideal. Will keep you posted.

**Comment by Nlammertyn** [comment]
I found the part where something goes wrong: https://github.com/CollectionFS/Meteor-cfs-s3/blob/master/s3.upload.stream2.js#L140

It's executed two times. Once when the next to last chunk has been uploaded and once when the last one has been uploaded. 

In other words uploadedSize === receivedSize when it shouldn't yet.

I'll go over the calculations of both tomorrow and should be able to work out the problem

**Comment by aldeed** [comment]
Awesome, thanks @Nlammertyn!

**Comment by zda** [comment]
Kudos @Nlammertyn. FWIW I just made an awful grody hack to ignore the initial execution on large files, and it fixes it for me (so I have faith that if you figure out the underlying double execution issue that'll solve the problem).

**Comment by raix** [comment]
@Nlammertyn cool :+1:

**Comment by Nlammertyn** [comment]
Fixed it, I'l fork and make a pull request now.

**Comment by Nlammertyn** [comment]
For those interested: https://github.com/CollectionFS/Meteor-cfs-s3/pull/44

**Comment by Nlammertyn** [comment]
This also resolves issues #243 (definitely the best of the 3 now), #254, #464 and #481
